File: python/agent_disclosure/attestation.py
"""ed25519 signature verification + freshness. Port of `src/attestation.ts`.

Asymmetric verification: the raw 32-byte public key travels as hex in the
envelope and is bound to `agentId`. A disclosure must be signed by the key it
claims as its identity — bound by a direct hex match, the `did:key` encoding of
that key (self-certifying form), or a verified rotation chain (`verify_key_binding`).

Two envelope shapes are supported, discriminated by shape:
  - v1 object envelope: `{disclosure, signature}` — `verify_disclosure_signature`.
  - v2 flattened JWS (EdDSA): `{payload, protected, header, signature}` —
    `verify_disclosure_jws`. The protected header (carrying `alg`) is part of the
    signed input, closing the v1 gap where the algorithm field sat outside the bytes.
"""

import base64
import json

from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.exceptions import InvalidSignature

from .canonical import canonicalize

# Multicodec prefix for an ed25519 public key (varint 0xed 0x01); did:key for ed25519
# is "did:key:z" + base58btc(0xed01 || rawPubKey). Mirrors src/did.ts.
_MULTICODEC_ED25519_PUB = bytes([0xED, 0x01])
_BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"


def _b64u_decode(s: str) -> bytes:
    """base64url-decode, restoring the stripped `=` padding (JWS uses unpadded form)."""
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def _b64u_encode(b: bytes) -> str:
    """base64url-encode without padding (the JWS on-wire form)."""
    return base64.urlsafe_b64encode(b).rstrip(b"=").decode("ascii")


def _base58_encode(data: bytes) -> str:
    """base58btc (Bitcoin alphabet) encode; leading zero bytes map to leading '1's."""
    zeros = 0
    while zeros < len(data) and data[zeros] == 0:
        zeros += 1
    num = int.from_bytes(data, "big")
    digits = ""
    while num > 0:
        num, rem = divmod(num, 58)
        digits = _BASE58_ALPHABET[rem] + digits
    return "1" * zeros + digits


def _agent_id_to_did_key(agent_id_hex: str) -> str:
    """Express an agentId (raw 32-byte ed25519 public key, hex) as a did:key.
    Mirrors `agentIdToDidKey` in src/did.ts. Raises on a non-32-byte key."""
    raw = bytes.fromhex(agent_id_hex)
    if len(raw) != 32:
        raise ValueError("agentId must be a 32-byte ed25519 public key (64 hex chars)")
    return "did:key:z" + _base58_encode(_MULTICODEC_ED25519_PUB + raw)


def verify_message(message: str, public_key_hex: str, signature: bytes) -> bool:
    """Verify an ed25519 `signature` (raw bytes) over a UTF-8 message against a
    32-byte raw public key (hex). Returns False on any malformed input."""
    try:
        raw = bytes.fromhex(public_key_hex)
        if len(raw) != 32:
            return False
        key = ed25519.Ed25519PublicKey.from_public_bytes(raw)
        key.verify(signature, message.encode("utf-8"))
        return True
    except (InvalidSignature, ValueError):
        return False


def verify_message_hex(message: str, public_key_hex: str, signature_hex: str) -> bool:
    """Verify a hex ed25519 signature over a UTF-8 message (v1 + handshake form)."""
    try:
        return verify_message(message, public_key_hex, bytes.fromhex(signature_hex))
    except ValueError:
        return False


def verify_key_binding(agent_id: str, signing_key_hex: str, rotation_chain=None):
    """Bind a disclosure's stable `agentId` to the key that actually signed it: a
    direct hex match, the did:key encoding of that key, or a verified rotation
    chain back to the agentId. Returns (ok, reason). Mirrors `verifyKeyBinding`."""
    if agent_id == signing_key_hex:
        return True, None
    try:
        if agent_id == _agent_id_to_did_key(signing_key_hex):
            return True, None
    except ValueError:
        # signing_key_hex is a valid 32-byte key on every call path here; defensive.
        pass
    if rotation_chain:
        return _verify_rotation_chain(agent_id, signing_key_hex, rotation_chain)
    return False, "agentId does not match the signing public key"


_MAX_ROTATION_CHAIN = 32


def _rotation_statement_body(frm: str, to: str, rotated_at: str) -> str:
    return canonicalize({"type": "rotation", "from": frm, "to": to, "rotatedAt": rotated_at})


def _verify_rotation_chain(agent_id: str, signing_key: str, chain):
    """Mirror `verifyRotationChain`: each hop's `from` must sign the move to its
    `to`; hops must be contiguous, acyclic, and end at the signing key."""
    if len(chain) == 0:
        return False, "empty rotation chain"
    if len(chain) > _MAX_ROTATION_CHAIN:
        return False, "rotation chain exceeds maximum length"
    cursor = agent_id
    seen = {agent_id}
    for s in chain:
        if s["from"] != cursor:
            return False, "rotation chain is not contiguous from agentId"
        if not verify_message_hex(
            _rotation_statement_body(s["from"], s["to"], s["rotatedAt"]), s["from"], s["signature"]
        ):
            return False, "a rotation statement signature does not verify against its from key"
        if s["to"] in seen:
            return False, "rotation chain contains a cycle"
        seen.add(s["to"])
        cursor = s["to"]
    if cursor != signing_key:
        return False, "rotation chain does not end at the signing key"
    return True, None


def verify_disclosure_signature(signed: dict):
    """Verify the ed25519 signature over the disclosure (v1 object envelope) and
    the agentId<->key binding. Returns (ok, reason); reason is None on success."""
    disclosure = signed["disclosure"]
    signature = signed["signature"]
    if not verify_message_hex(canonicalize(disclosure), signature["publicKey"], signature["value"]):
        return False, "signature mismatch"
    return verify_key_binding(disclosure["agentId"], signature["publicKey"], signed.get("rotationChain"))


def verify_disclosure_jws(signed: dict):
    """Verify a v2 JWS envelope: the protected header must declare EdDSA, the
    signature must verify over ASCII(protected + "." + payload) against the JWK
    key, and the payload's `agentId` must bind to that key. Mirrors
    `verifyDisclosureJws`. Returns (ok, reason)."""
    try:
        header = json.loads(_b64u_decode(signed["protected"]).decode("utf-8"))
    except Exception:  # noqa: BLE001 — any decode fault is a safe rejection
        return False, "unreadable protected header"
    if header.get("alg") != "EdDSA":
        return False, f"unsupported JWS alg: {header.get('alg')}"

    try:
        pub_hex = _b64u_decode(signed["header"]["jwk"]["x"]).hex()
    except Exception:  # noqa: BLE001
        return False, "unreadable jwk.x"
    if len(pub_hex) != 64:
        return False, "jwk.x is not a 32-byte ed25519 key"

    signing_input = f"{signed['protected']}.{signed['payload']}"
    try:
        sig = _b64u_decode(signed["signature"])
    except Exception:  # noqa: BLE001
        return False, "unreadable signature"
    if not verify_message(signing_input, pub_hex, sig):
        return False, "jws signature mismatch"

    try:
        agent_id = json.loads(_b64u_decode(signed["payload"]).decode("utf-8")).get("agentId")
    except Exception:  # noqa: BLE001
        return False, "unreadable payload"
    if not isinstance(agent_id, str):
        return False, "payload has no agentId"
    return verify_key_binding(agent_id, pub_hex, signed.get("rotationChain"))


def is_jws_signed_disclosure(signed) -> bool:
    """True if the envelope is the v2 flattened-JWS form (discriminated by shape)."""
    return (
        isinstance(signed, dict)
        and isinstance(signed.get("payload"), str)
        and isinstance(signed.get("protected"), str)
    )


def verify_any_disclosure_signature(signed: dict):
    """Verify either envelope shape (v1 object or v2 flattened JWS)."""
    if is_jws_signed_disclosure(signed):
        return verify_disclosure_jws(signed)
    return verify_disclosure_signature(signed)


def disclosure_of(signed: dict) -> dict:
    """Extract the disclosure document from either envelope shape. For v2 it
    base64url-decodes the JCS payload."""
    if is_jws_signed_disclosure(signed):
        return json.loads(_b64u_decode(signed["payload"]).decode("utf-8"))
    return signed["disclosure"]


def is_fresh(disclosure: dict, now: str) -> bool:
    """A disclosure is valid only within [issuedAt, validUntil]. ISO-8601
    timestamps compare lexically, matching the TS string comparison."""
    return disclosure["issuedAt"] <= now <= disclosure["validUntil"]

File: src/attestation.ts
// Attestation primitives - sign a disclosure so it resists post-hoc rewriting, and
// let a COUNTERPARTY verify it with no shared secret. ed25519 (asymmetric): the
// signer holds the private key; the public key travels in the envelope and is
// bound to the agent's id. Vendor-neutral (node:crypto only) so it lifts into the
// standalone repo with the schema.

import {
  createHash,
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  sign as edSign,
  verify as edVerify,
  type KeyObject,
} from "node:crypto";
import { agentIdToDidKey } from "./did.ts";
import type { AgentDisclosure, JwsSignedDisclosure, SignedDisclosure } from "./schema.ts";
// Type-only (erased at runtime) so attestation stays the lowest layer: keys.ts
// imports the shared body-builder below, this only borrows the shape. No runtime cycle.
import type { RotationStatement } from "./keys.ts";

// SPKI DER prefix for an ed25519 public key; prepended to the raw 32-byte key so
// Node can import a bare hex public key (the interoperable on-wire form).
const SPKI_ED25519_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export interface AgentKeyPair {
  privateKey: KeyObject;
  publicKey: KeyObject;
  /** raw 32-byte public key as hex - this is the agentId + envelope publicKey */
  publicKeyHex: string;
}

/** Mint a fresh agent signing identity. */
export function generateAgentKeyPair(): AgentKeyPair {
  const { privateKey, publicKey } = generateKeyPairSync("ed25519");
  const der = publicKey.export({ type: "spki", format: "der" });
  const raw = der.subarray(der.length - 32);
  return { privateKey, publicKey, publicKeyHex: raw.toString("hex") };
}

function publicKeyFromHex(hex: string): KeyObject {
  const raw = Buffer.from(hex, "hex");
  if (raw.length !== 32) throw new Error("ed25519 public key must be 32 bytes");
  return createPublicKey({ key: Buffer.concat([SPKI_ED25519_PREFIX, raw]), format: "der", type: "spki" });
}

/** Sign an arbitrary UTF-8 message with the agent key (hex signature). The generic
 *  primitive the disclosure signing + the challenge handshake both build on. */
export function signMessage(message: string, key: AgentKeyPair): string {
  return edSign(null, Buffer.from(message, "utf8"), key.privateKey).toString("hex");
}

/** Verify a hex signature over a UTF-8 message against an ed25519 public key (hex). */
export function verifyMessage(message: string, publicKeyHex: string, signatureHex: string): boolean {
  try {
    return edVerify(null, Buffer.from(message, "utf8"), publicKeyFromHex(publicKeyHex), Buffer.from(signatureHex, "hex"));
  } catch {
    return false;
  }
}

/** Serialize the private key (PKCS8 DER hex) so an agent's signing identity is
 *  stable across restarts. Pair with `agentKeyFromPrivateHex`. */
export function exportAgentKey(key: AgentKeyPair): string {
  return (key.privateKey.export({ type: "pkcs8", format: "der" }) as Buffer).toString("hex");
}

/** Reconstruct a full key pair from a persisted private key (PKCS8 DER hex). */
export function agentKeyFromPrivateHex(hex: string): AgentKeyPair {
  const privateKey = createPrivateKey({ key: Buffer.from(hex, "hex"), format: "der", type: "pkcs8" });
  const publicKey = createPublicKey(privateKey);
  const der = publicKey.export({ type: "spki", format: "der" });
  return { privateKey, publicKey, publicKeyHex: der.subarray(der.length - 32).toString("hex") };
}

/** Defense-in-depth: cap recursion so a hostile, deeply-nested value cannot exhaust
 *  the stack. No valid disclosure nests anywhere near this; the byte output for any
 *  in-range value is unchanged. */
export const MAX_CANONICALIZE_DEPTH = 256;

/** Deterministic JSON: keys sorted recursively, so the signed bytes are stable
 *  across producers (the same canonicalization the audit chain uses). */
export function canonicalize(value: unknown): string {
  return canonicalizeAt(value, 0);
}

function canonicalizeAt(value: unknown, depth: number): string {
  if (depth > MAX_CANONICALIZE_DEPTH) throw new Error("canonicalize: maximum nesting depth exceeded");
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map((v) => canonicalizeAt(v, depth + 1)).join(",")}]`;
  const obj = value as Record<string, unknown>;
  const body = Object.keys(obj)
    .sort()
    .filter((k) => obj[k] !== undefined)
    .map((k) => `${JSON.stringify(k)}:${canonicalizeAt(obj[k], depth + 1)}`)
    .join(",");
  return `{${body}}`;
}

/** sha256 hex of a string - used for the various digest/fingerprint fields. */
export function sha256Hex(input: string): string {
  return createHash("sha256").update(input).digest("hex");
}

/** Sign a disclosure with an agent key, returning the signed envelope. */
export function signDisclosure(disclosure: AgentDisclosure, key: AgentKeyPair): SignedDisclosure {
  return {
    disclosure,
    signature: { algorithm: "ed25519", publicKey: key.publicKeyHex, value: signMessage(canonicalize(disclosure), key) },
  };
}

export interface SignatureCheck {
  ok: boolean;
  reason?: string;
}

/** Canonical signed body of a key-rotation statement. Single source of truth shared
 *  with keys.ts (which imports it) so the signing side and the chain-verifying side
 *  here can never disagree on the bytes. */
export function rotationStatementBody(from: string, to: string, rotatedAt: string): string {
  return canonicalize({ type: "rotation", from, to, rotatedAt });
}

/** Defense: cap a rotation chain so a hostile envelope can't force unbounded work. */
export const MAX_ROTATION_CHAIN = 32;

/** Verify that a chain of signed rotation statements links the stable `agentId` to the
 *  key that actually signed a disclosure (`signingKey`). Each hop's `from` must itself
 *  sign the move to its `to`, the hops must be contiguous, acyclic, and end at the
 *  signing key. This is what lets an identity survive key rotation: `agentId` stays
 *  fixed while the signing key advances. */
export function verifyRotationChain(
  agentId: string,
  signingKey: string,
  chain: readonly RotationStatement[],
): SignatureCheck {
  if (chain.length === 0) return { ok: false, reason: "empty rotation chain" };
  if (chain.length > MAX_ROTATION_CHAIN) return { ok: false, reason: "rotation chain exceeds maximum length" };
  let cursor = agentId;
  const seen = new Set<string>([agentId]);
  for (const s of chain) {
    if (s.from !== cursor) return { ok: false, reason: "rotation chain is not contiguous from agentId" };
    if (!verifyMessage(rotationStatementBody(s.from, s.to, s.rotatedAt), s.from, s.signature)) {
      return { ok: false, reason: "a rotation statement signature does not verify against its from key" };
    }
    if (seen.has(s.to)) return { ok: false, reason: "rotation chain contains a cycle" };
    seen.add(s.to);
    cursor = s.to;
  }
  if (cursor !== signingKey) return { ok: false, reason: "rotation chain does not end at the signing key" };
  return { ok: true };
}

/** Bind a disclosure's stable `agentId` to the key that actually signed it: a direct hex
 *  match, the did:key encoding of that key (self-certifying form), or a verified rotation
 *  chain back to the agentId. Shared by both envelope shapes. */
export function verifyKeyBinding(
  agentId: string,
  signingKeyHex: string,
  rotationChain?: readonly RotationStatement[],
): SignatureCheck {
  if (agentId === signingKeyHex) return { ok: true };
  try {
    if (agentId === agentIdToDidKey(signingKeyHex)) return { ok: true };
  } catch {
    // signingKeyHex is a valid 32-byte key on every call path here; defensive only.
  }
  if (rotationChain?.length) return verifyRotationChain(agentId, signingKeyHex, rotationChain);
  return { ok: false, reason: "agentId does not match the signing public key" };
}

/** Verify the ed25519 signature over the disclosure (v1 object envelope). Pure; no policy
 *  applied here (see verify.ts for the counterparty decision). Also enforces the
 *  agentId↔key binding via `verifyKeyBinding`. */
export function verifyDisclosureSignature(signed: SignedDisclosure): SignatureCheck {
  if (!verifyMessage(canonicalize(signed.disclosure), signed.signature.publicKey, signed.signature.value)) {
    return { ok: false, reason: "signature mismatch" };
  }
  return verifyKeyBinding(signed.disclosure.agentId, signed.signature.publicKey, signed.rotationChain);
}

// ── v2: flattened JWS (EdDSA) envelope ───────────────────────────────────────
const JWS_PROTECTED_HEADER = { alg: "EdDSA", typ: "application/adp+json" } as const;

/** Sign a disclosure as a flattened JWS (EdDSA) envelope — the JOSE-interoperable v2
 *  wrapping. The signature covers ASCII(b64u(protected) + "." + b64u(payload)), so the
 *  protected header (carrying `alg`) is integrity-protected — closing the v1 gap where
 *  the algorithm field sat outside the signed bytes. Payload is the same RFC 8785 (JCS)
 *  canonical document, so a JOSE library can verify it. */
export function signDisclosureJws(disclosure: AgentDisclosure, key: AgentKeyPair): JwsSignedDisclosure {
  const protectedB64 = Buffer.from(JSON.stringify(JWS_PROTECTED_HEADER), "utf8").toString("base64url");
  const payloadB64 = Buffer.from(canonicalize(disclosure), "utf8").toString("base64url");
  const signature = edSign(null, Buffer.from(`${protectedB64}.${payloadB64}`, "ascii"), key.privateKey).toString("base64url");
  return {
    payload: payloadB64,
    protected: protectedB64,
    header: { jwk: { kty: "OKP", crv: "Ed25519", x: Buffer.from(key.publicKeyHex, "hex").toString("base64url") } },
    signature,
  };
}

/** Verify a v2 JWS envelope: the protected header must declare EdDSA, the signature must
 *  verify over the signing input against the JWK key, and the payload's `agentId` must
 *  bind to that key (direct, did:key, or rotation chain). */
export function verifyDisclosureJws(signed: JwsSignedDisclosure): SignatureCheck {
  let header: { alg?: unknown };
  try {
    header = JSON.parse(Buffer.from(signed.protected, "base64url").toString("utf8"));
  } catch {
    return { ok: false, reason: "unreadable protected header" };
  }
  if (header.alg !== "EdDSA") return { ok: false, reason: `unsupported JWS alg: ${String(header.alg)}` };

  const pubHex = Buffer.from(signed.header.jwk.x, "base64url").toString("hex");
  if (pubHex.length !== 64) return { ok: false, reason: "jwk.x is not a 32-byte ed25519 key" };

  let okSig = false;
  try {
    okSig = edVerify(
      null,
      Buffer.from(`${signed.protected}.${signed.payload}`, "ascii"),
      publicKeyFromHex(pubHex),
      Buffer.from(signed.signature, "base64url"),
    );
  } catch {
    return { ok: false, reason: "jws signature mismatch" };
  }
  if (!okSig) return { ok: false, reason: "jws signature mismatch" };

  let agentId: unknown;
  try {
    agentId = (JSON.parse(Buffer.from(signed.payload, "base64url").toString("utf8")) as { agentId?: unknown }).agentId;
  } catch {
    return { ok: false, reason: "unreadable payload" };
  }
  if (typeof agentId !== "string") return { ok: false, reason: "payload has no agentId" };
  return verifyKeyBinding(agentId, pubHex, signed.rotationChain);
}

/** Verify either envelope shape (v1 object or v2 flattened JWS), discriminated by shape. */
export function verifyAnyDisclosureSignature(signed: SignedDisclosure | JwsSignedDisclosure): SignatureCheck {
  return "payload" in signed && "protected" in signed ? verifyDisclosureJws(signed) : verifyDisclosureSignature(signed);
}

/** Freshness: a disclosure is valid only within [issuedAt, validUntil]. ISO-8601
 *  timestamps compare lexically, so string comparison is correct here. */
export function isFresh(disclosure: AgentDisclosure, now: string): boolean {
  return now >= disclosure.issuedAt && now <= disclosure.validUntil;
}
